File: forge-core/internal/planningownership/yaml_flow.go
package planningownership

import "strings"

type flowCursor struct {
	text   string
	index  int
	parser *yamlParser
}

func (parser *yamlParser) parseInlineValue(text string, depth int) (any, error) {
	if depth > maxYAMLDepth {
		return nil, parser.errorf("YAML depth exceeds %d", maxYAMLDepth)
	}
	if text[0] != '[' && text[0] != '{' && text[0] != '"' {
		return parser.parsePlainScalar(text)
	}
	cursor := &flowCursor{text: text, parser: parser}
	value, err := cursor.parseValue(depth)
	if err != nil {
		return nil, err
	}
	cursor.skipSpaces()
	if cursor.index != len(cursor.text) {
		return nil, parser.errorf("trailing inline YAML content")
	}
	return value, nil
}

func (cursor *flowCursor) parseValue(depth int) (any, error) {
	if depth > maxYAMLDepth {
		return nil, cursor.parser.errorf("YAML depth exceeds %d", maxYAMLDepth)
	}
	cursor.skipSpaces()
	if cursor.index >= len(cursor.text) {
		return nil, cursor.parser.errorf("missing inline YAML value")
	}
	switch cursor.text[cursor.index] {
	case '[':
		return cursor.parseSequence(depth)
	case '{':
		return cursor.parseMapping(depth)
	case '"':
		return cursor.parseQuoted()
	default:
		return cursor.parsePlain()
	}
}

func (cursor *flowCursor) parseSequence(depth int) ([]any, error) {
	if err := cursor.parser.addCollection(depth); err != nil {
		return nil, err
	}
	cursor.index++
	cursor.skipSpaces()
	result := make([]any, 0)
	if cursor.consume(']') {
		return result, nil
	}
	for {
		if len(result) >= maxYAMLItems {
			return nil, cursor.parser.errorf("flow sequence exceeds %d items", maxYAMLItems)
		}
		value, err := cursor.parseValue(depth + 1)
		if err != nil {
			return nil, err
		}
		result = append(result, value)
		cursor.skipSpaces()
		if cursor.consume(']') {
			return result, nil
		}
		if !cursor.consume(',') {
			return nil, cursor.parser.errorf("flow sequence requires comma or closing bracket")
		}
		cursor.skipSpaces()
		if cursor.peek(']') {
			return nil, cursor.parser.errorf("flow sequence trailing comma is forbidden")
		}
	}
}

func (cursor *flowCursor) parseMapping(depth int) (map[string]any, error) {
	if err := cursor.parser.addCollection(depth); err != nil {
		return nil, err
	}
	cursor.index++
	cursor.skipSpaces()
	result := make(map[string]any)
	if cursor.consume('}') {
		return result, nil
	}
	for {
		if len(result) >= maxYAMLFields {
			return nil, cursor.parser.errorf("flow mapping exceeds %d fields", maxYAMLFields)
		}
		if err := cursor.parseMappingEntry(result, depth); err != nil {
			return nil, err
		}
		cursor.skipSpaces()
		if cursor.consume('}') {
			return result, nil
		}
		if !cursor.consume(',') {
			return nil, cursor.parser.errorf("flow mapping requires comma or closing brace")
		}
		cursor.skipSpaces()
		if cursor.peek('}') {
			return nil, cursor.parser.errorf("flow mapping trailing comma is forbidden")
		}
	}
}

func (cursor *flowCursor) parseMappingEntry(result map[string]any, depth int) error {
	key, err := cursor.parseKey()
	if err != nil {
		return err
	}
	if !cursor.consume(':') {
		return cursor.parser.errorf("flow mapping key lacks colon")
	}
	if cursor.index >= len(cursor.text) || cursor.text[cursor.index] != ' ' ||
		cursor.index+1 >= len(cursor.text) || cursor.text[cursor.index+1] == ' ' {
		return cursor.parser.errorf("flow mapping colon must be followed by exactly one space")
	}
	cursor.index++
	if _, duplicate := result[key]; duplicate {
		return cursor.parser.errorf("duplicate flow mapping key %q", key)
	}
	if err := cursor.parser.addToken(); err != nil {
		return err
	}
	value, err := cursor.parseValue(depth + 1)
	if err != nil {
		return err
	}
	result[key] = value
	return nil
}

func (cursor *flowCursor) parseKey() (string, error) {
	cursor.skipSpaces()
	if cursor.peek('"') {
		key, err := cursor.parseQuotedText()
		if err != nil {
			return "", err
		}
		if !validYAMLKey(key) || key == "<<" {
			return "", cursor.parser.errorf("invalid or forbidden flow mapping key %q", key)
		}
		return key, nil
	}
	start := cursor.index
	for cursor.index < len(cursor.text) && cursor.text[cursor.index] != ':' {
		if strings.ContainsRune("{},[]", rune(cursor.text[cursor.index])) {
			return "", cursor.parser.errorf("flow mapping key is malformed")
		}
		cursor.index++
	}
	key := cursor.text[start:cursor.index]
	if !validYAMLKey(key) || key == "<<" {
		return "", cursor.parser.errorf("invalid or forbidden flow mapping key %q", key)
	}
	return key, nil
}

func (cursor *flowCursor) parseQuoted() (any, error) {
	value, err := cursor.parseQuotedText()
	if err != nil {
		return nil, err
	}
	if err := cursor.parser.addToken(); err != nil {
		return nil, err
	}
	return value, nil
}

func (cursor *flowCursor) parseQuotedText() (string, error) {
	cursor.index++
	start := cursor.index
	for cursor.index < len(cursor.text) && cursor.text[cursor.index] != '"' {
		if cursor.text[cursor.index] == '\\' {
			return "", cursor.parser.errorf("escape bytes are forbidden in double-quoted YAML")
		}
		cursor.index++
	}
	if cursor.index >= len(cursor.text) {
		return "", cursor.parser.errorf("unterminated double-quoted YAML scalar")
	}
	value := cursor.text[start:cursor.index]
	cursor.index++
	if len(value) > maxYAMLScalarBytes {
		return "", cursor.parser.errorf("quoted scalar exceeds %d bytes", maxYAMLScalarBytes)
	}
	return value, nil
}

func (cursor *flowCursor) parsePlain() (any, error) {
	start := cursor.index
	for cursor.index < len(cursor.text) && !strings.ContainsRune(",]}", rune(cursor.text[cursor.index])) {
		cursor.index++
	}
	return cursor.parser.parsePlainScalar(cursor.text[start:cursor.index])
}

func (cursor *flowCursor) skipSpaces() {
	for cursor.index < len(cursor.text) && cursor.text[cursor.index] == ' ' {
		cursor.index++
	}
}

func (cursor *flowCursor) consume(expected byte) bool {
	if cursor.index >= len(cursor.text) || cursor.text[cursor.index] != expected {
		return false
	}
	cursor.index++
	return true
}

func (cursor *flowCursor) peek(expected byte) bool {
	return cursor.index < len(cursor.text) && cursor.text[cursor.index] == expected
}

File: harness/planning_capability_ownership_projection/yaml_flow.py
"""Recursive parser for the frozen flow-map and flow-sequence subset."""

from __future__ import annotations

from .codec import ContractError
from .yaml_resources import Resources, bounded_mapping, bounded_sequence
from .yaml_scalars import mapping_key, scalar


class FlowParser:
    def __init__(self, text: str, resources: Resources, depth: int):
        self.text, self.resources, self.depth, self.at = text, resources, depth, 0

    def parse(self) -> object:
        value = self._value(self.depth)
        self._spaces()
        if self.at != len(self.text):
            raise ContractError("trailing YAML flow content")
        return value

    def _spaces(self) -> None:
        while self.at < len(self.text) and self.text[self.at] == " ":
            self.at += 1

    def _take(self, expected: str) -> None:
        self._spaces()
        if self.at >= len(self.text) or self.text[self.at] != expected:
            raise ContractError(f"expected YAML flow token {expected!r}")
        self.at += 1

    def _value(self, depth: int) -> object:
        self.resources.depth(depth)
        self._spaces()
        if self.at >= len(self.text):
            raise ContractError("missing YAML flow value")
        if self.text[self.at] == "[":
            return self._sequence(depth)
        if self.text[self.at] == "{":
            return self._mapping(depth)
        return scalar(self._scalar_text(",]}"), self.resources)

    def _scalar_text(self, delimiters: str) -> str:
        start, quoted = self.at, False
        while self.at < len(self.text):
            character = self.text[self.at]
            if character == '"':
                quoted = not quoted
            if not quoted and character in delimiters:
                break
            self.at += 1
        if quoted:
            raise ContractError("unterminated YAML flow quote")
        return self.text[start:self.at].strip()

    def _sequence(self, depth: int) -> list[object]:
        self.resources.collection(depth)
        self._take("[")
        values: list[object] = []
        self._spaces()
        if self.at < len(self.text) and self.text[self.at] == "]":
            self.at += 1
            return values
        while True:
            values.append(self._value(depth + 1))
            bounded_sequence(values)
            self._spaces()
            if self.at < len(self.text) and self.text[self.at] == "]":
                self.at += 1
                return values
            self._take(",")
            self._spaces()
            if self.at < len(self.text) and self.text[self.at] == "]":
                raise ContractError("trailing YAML flow comma forbidden")

    def _flow_key(self) -> str:
        self._spaces()
        start = self.at
        while self.at < len(self.text) and self.text[self.at] not in ":,{}[]":
            self.at += 1
        raw = self.text[start:self.at]
        if raw != raw.strip():
            raise ContractError("YAML flow mapping key has padding")
        return mapping_key(raw, self.resources)

    def _mapping(self, depth: int) -> dict[str, object]:
        self.resources.collection(depth)
        self._take("{")
        result: dict[str, object] = {}
        self._spaces()
        if self.at < len(self.text) and self.text[self.at] == "}":
            self.at += 1
            return result
        while True:
            key = self._flow_key()
            if key in result:
                raise ContractError(f"duplicate YAML key {key!r}")
            if self.at >= len(self.text) or self.text[self.at] != ":":
                raise ContractError("missing YAML flow mapping colon")
            self.at += 1
            if (self.at + 1 >= len(self.text) or self.text[self.at] != " " or
                    self.text[self.at + 1] == " "):
                raise ContractError("YAML flow colon requires exactly one space and a value")
            self.at += 1
            result[key] = self._value(depth + 1)
            bounded_mapping(result)
            self._spaces()
            if self.at < len(self.text) and self.text[self.at] == "}":
                self.at += 1
                return result
            self._take(",")
            self._spaces()
            if self.at < len(self.text) and self.text[self.at] == "}":
                raise ContractError("trailing YAML flow comma forbidden")


def parse_flow(text: str, resources: Resources, depth: int) -> object:
    return FlowParser(text, resources, depth).parse()
